File: src/commands/announcements.ts
import { getOutputFormat, formatTimestamp } from "../lib/utils.ts";
import { Command } from "commander";
import type { OutputFormat } from "../lib/types.ts";
import { getSiteInfoApi, getMessagesApi, getDiscussionPostsApi } from "../lib/moodle.ts";
import { createApiContext } from "../lib/auth.ts";
import { formatAndOutput } from "../index.ts";

export function registerAnnouncementsCommand(program: Command): void {
  const announcementsCmd = program.command("announcements");
  announcementsCmd.description("Announcement operations");

  announcementsCmd
    .command("list-all")
    .description("List all announcements across all courses")
    .option("--level <type>", "Course level: in_progress (default) | all", "in_progress")
    .option("--unread-only", "Show only unread announcements")
    .option("--limit <n>", "Maximum number of announcements to show", "20")
    .option("--output <format>", "Output format: json|csv|table|silent")
    .action(async (options, command) => {
      const output: OutputFormat = getOutputFormat(command);
      const limit = parseInt(options.limit, 10);
      const apiContext = await createApiContext(options, command);
      if (!apiContext) {
        process.exitCode = 1;
        return;
      }

      const siteInfo = await getSiteInfoApi(apiContext.session);

      const messages = await getMessagesApi(apiContext.session, siteInfo.userid, {
        limitnum: limit,
      });

      const allAnnouncements = messages.map(m => ({
        course_id: 0,
        course_name: "Notifications",
        id: m.id,
        subject: m.subject,
        author: `User ${m.useridfrom}`,
        author_id: m.useridfrom,
        created_at: formatTimestamp(m.timecreated),
        modified_at: formatTimestamp(m.timecreated),
        unread: false,
      }));

      allAnnouncements.sort((a, b) => (b as any).created_at > (a as any).created_at ? 1 : -1);

      const shown = allAnnouncements.slice(0, limit);

      formatAndOutput(
        shown as unknown as Record<string, unknown>[],
        output,
        apiContext.log,
        { status: "success", timestamp: new Date().toISOString(), total_announcements: allAnnouncements.length, shown: shown.length }
      );
    });

  announcementsCmd
    .command("read")
    .description("Read a specific announcement (shows full content)")
    .argument("<announcement-id>", "Discussion ID of the announcement")
    .option("--output <format>", "Output format: json|csv|table|silent")
    .action(async (announcementId, options, command) => {
      const output: OutputFormat = getOutputFormat(command);
      const apiContext = await createApiContext(options, command);
      if (!apiContext) {
        process.exitCode = 1;
        return;
      }

      const posts = await getDiscussionPostsApi(apiContext.session, parseInt(announcementId, 10));

      if (posts.length === 0) {
        apiContext.log.error(`Announcement not found: ${announcementId}`);
        process.exitCode = 1;
        return;
      }

      const firstPost = posts[0];

      formatAndOutput(
        {
          id: announcementId,
          subject: firstPost.subject,
          author: firstPost.author,
          author_id: firstPost.authorId,
          created_at: formatTimestamp(firstPost.created),
          modified_at: formatTimestamp(firstPost.modified),
          message: firstPost.message,
        } as unknown as Record<string, unknown>,
        output,
        apiContext.log
      );
    });
}

File: src/commands/announcements.rs
use super::ApiCtx;
use crate::moodle::message::get_messages_api;
use crate::output::format_and_output;
use crate::utils::format_moodle_date;
use crate::Cli;
use anyhow::Result;

pub async fn run(cmd: &crate::AnnouncementsCommands, cli: &Cli) -> Result<()> {
    let ctx = ApiCtx::build(cli)?;

    match cmd {
        crate::AnnouncementsCommands::ListAll { unread_only, limit } => {
            let messages = get_messages_api(
                &ctx.client,
                &ctx.session,
                ctx.session.user_id,
                None,
                if *unread_only { Some(false) } else { None },
                Some(*limit),
            )
            .await?;

            let mut items: Vec<serde_json::Value> = messages
                .iter()
                .map(|m| {
                    serde_json::json!({
                        "course_id": 0,
                        "course_name": "Notifications",
                        "id": m.id,
                        "subject": m.subject,
                        "author": format!("User {}", m.useridfrom),
                        "author_id": m.useridfrom,
                        "created_at": format_moodle_date(Some(m.timecreated)),
                        "unread": !m.read,
                    })
                })
                .collect();

            // Sort by created_at descending (already strings, lexicographic sort works for ISO dates)
            items.sort_by(|a, b| {
                let ta = a.get("created_at").and_then(|v| v.as_str()).unwrap_or("");
                let tb = b.get("created_at").and_then(|v| v.as_str()).unwrap_or("");
                tb.cmp(ta)
            });

            let shown: Vec<_> = items.into_iter().take(*limit as usize).collect();

            ctx.log
                .info(&format!("Showing {} announcements", shown.len()));
            format_and_output(&shown, ctx.output, None);
        }

        crate::AnnouncementsCommands::Read { announcement_id } => {
            let messages = get_messages_api(
                &ctx.client,
                &ctx.session,
                ctx.session.user_id,
                None,
                None,
                None,
            )
            .await?;
            let message = messages
                .iter()
                .find(|message| message.id == *announcement_id)
                .ok_or_else(|| anyhow::anyhow!("Announcement not found: {}", announcement_id))?;
            let item = serde_json::json!({
                "id": announcement_id,
                "subject": message.subject,
                "author": format!("User {}", message.useridfrom),
                "author_id": message.useridfrom,
                "created_at": format_moodle_date(Some(message.timecreated)),
                "message": message.text,
            });

            format_and_output(&[item], ctx.output, None);
        }
    }

    Ok(())
}
